gitrevise/merge.py

```diff
@@ -199,29 +202,131 @@ def merge_blobs(
         conflicts = tmpdir / "conflict" / path.relative_to("/")
         conflicts.parent.mkdir(parents=True, exist_ok=True)
-        conflicts.write_bytes(err.output)
+        conflicts.write_bytes(preimage)
         merged = edit_file(repo, conflicts)
 
         # Print warnings if the merge looks like it may have failed.
-        if merged == err.output:
+        if merged == preimage:
             print("(note) conflicted file is unchanged")
 
-        if b"<<<<<<<" in merged or b"=======" in merged or b">>>>>>>" in merged:
+        if (
+            b"<" * MARKER_SIZE in merged
+            or b"=" * MARKER_SIZE in merged
+            or b">" * MARKER_SIZE in merged
+        ):
             print("(note) conflict markers found in the merged file")
 
         # Was the merge successful?
         if input("  Merge successful? (y/N) ").lower() != "y":
-            raise MergeConflict("user aborted")
+            raise MergeConflict("user aborted")  # pylint: disable=W0707
+
+        record_resolution(repo, conflict_id, preimage, merged)
 
     return Blob(current.repo, merged)
+
+
+def reuse_recorded_resolution(repo, preimage) -> Tuple[Optional[Blob], Optional[str]]:
+    rr_cache = repo.gitdir / "rr-cache"
+    # TODO Should we cache settings like this?
+    if not repo.bool_config("rerere.enabled", default=rr_cache.is_dir()):
+        return None, None
+
+    conflict_id = conflict_id_by_file_contents(preimage)
+    conflict_dir = rr_cache / conflict_id
+    if not conflict_dir.is_dir():
+        return None, conflict_id
+    if not repo.bool_config("rerere.autoUpdate", default=False):
+        return None, conflict_id
+    postimage_path = conflict_dir / "postimage"
+    postimage = postimage_path.read_bytes()
+    recorded_preimage = (conflict_dir / "preimage").read_bytes()
+    try:
+        result = merge_blobs(
+            repo.workdir,  # TODO is this correct?
+            labels=(
+                "current",
+                "base",
+                "other",
+            ),
+            current=Blob(repo, postimage),
+            base=Blob(repo, recorded_preimage),
+            other=Blob(repo, preimage),
+        )
+    except MergeConflict:
+        # We could ask the user to merge this. However, that could be confusing.
+        # Just fall back to letting them resolve the entire conflict.
+        return None, None
+    print("Successfully replayed recorded resolution")
+    # Mark that "postimage" was used to help git gc (git does the same).
+    os.utime(postimage_path)
+    return result, conflict_id
+
+
+def record_resolution(
+    repo: Repository, conflict_id: Optional[str], preimage: bytes, postimage: bytes
+) -> None:
+    if conflict_id is None:
+        return
+
+    conflict_dir = repo.gitdir / "rr-cache" / conflict_id
+    conflict_dir.parent.mkdir(exist_ok=True)
+    conflict_dir.mkdir(exist_ok=True)
+    print("Recording conflict resolution")
+    (conflict_dir / "preimage").write_bytes(preimage)
```

Comment:
I think you need to normalize the conflict markers in this preimage before writing it to disk. 

It also _might_ be worth checking you're not overriding a file before writing it, but I imagine git will handle an overwritten preimage/postimage just fine.

Finally, we should probably catch exceptions thrown in this method, warn, and discard them, as we probably don't want to hard-fail if we fail to record a resolution.

Reply:
Currently I'm still overwriting the files which I'd assume is not a problem.It seems better than not overwriting an old resolution which might no longer be usable because it doesn't merge cleanly.

I need to look into the Git implementation a bit more. I plan to implement variants as well, then there will no longer be the need to overwrite old conflict data.